### PR TITLE
Kf tasks

### PR DESCRIPTION
**Pull request by sponnet**
- logs refactored
- tasks refactored
- taskinprogress set to false when all tasks ready ( or fail )

### CODE REVIEW

File: API/package.json [hunk 1]
```diff
@@ -10,7 +10,12 @@
   "license": "MIT",
   "dependencies": {
     "express": "^4.15.3",
+    "memory-cache": "^0.2.0",
```
Comment (bkawk) [inline API/package.json]: I really don't feel we need this dependency

File: API/package.json [hunk 2]
```diff
@@ -10,7 +10,12 @@
...
+    "request": "^2.83.0",
     "socket.io": "^2.0.3",
     "web3": "^1.0.0-beta.22"
+  },
+  "devDependencies": {
+    "should": "^13.1.2"
```
Comment (bkawk) [inline API/package.json]: Can this be used with Polymer too? I think we should select a testing librabry that can be used front and back

File: API/tasks/Avatar.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+module.exports = function(web3) {
+
+const ipfs = require('../js-ipfs-api.js');
```
Comment (bkawk) [inline API/tasks/Avatar.js]: This is all done in the /Store container, we just need to send the task to that comntainer

File: API/tasks/getBalance.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	return ({
+		/**
+		 * Get Balance
+		 * @param {String} address - The user's pub key
+		 * @param {Array} tokens - The tokens to be fetched
+		 * @return {Array} - Balances of that user
+		 */
+
+		/**
+		 * Represents a users balance
```
Comment (bkawk) [inline API/tasks/getBalance.js]: this js doc only needs to detail the getBalance function

File: API/tasks/getFx.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+						let p = JSON.parse(body);
+
+						let r = {
+							price_btc: p[0].price_btc,
+							price_eur: p[0].price_eur,
+							price_usd: p[0].price_usd,
+							symbol: p[0].symbol
+						};
+
+            // we should cache the response, to avoid our IP
```
Comment (bkawk) [inline API/tasks/getFx.js]: We should reschedule this in the queue to avoid this issue

File: API/tasks/getGasPrice.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+module.exports = function(web3) {
+
+	return ({
+
+		_getGasPrice: function(address) {
+
+			return new Promise((resolve, reject) => {
+				web3.eth.getGasPrice().then((gasPrice) => {
+					console.log('GasPrice: ', gasPrice.toString(10));
```
Comment (bkawk) [inline API/tasks/getGasPrice.js]: remove log

File: API/tasks/getHashtagItems.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
+module.exports = function(web3) {
+
+	return ({
+
+		_getHashtagItems: function(hashtag) {
+
+
+			/**
```
Comment (bkawk) [inline API/tasks/getHashtagItems.js]: the jsdoc should be above the function not in and only detail the arguments and returns of the function

File: Site/index.html [hunk 8]
```diff
@@ -60,13 +60,14 @@
...
     <script src='socket.io.js'></script>
     <script>
-      const socket = io.connect('http://localhost:8011', {query: 'publicKey=0x0345345222222234534kj5h3kj45h'});
+      var publicKey = localStorage.getItem('publickeytest');
+      const socket = io.connect('http://localhost:8011', {query: 'publicKey='+publicKey});
 
-      socket.emit('saveError', 'error!!', function (data) {
-        console.log(data); 
+      socket.emit('createItem', 'DEALDATA', function (data) {
+        console.log(data);
```
Comment (bkawk) [inline Site/index.html]: remove log

File: API/tasks/getHashtags.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+				hashtagItems: 31
+			}, {
+				contractAddress: "0x000",
+				hashtagName: "SponnetShop",
+				hashtagBalance: "102",
+				hashtagItems: 11
+			}];
+
+			return new Promise((resolve, reject) => {
+				console.log('Hashtags: ', hashtags);
```
Comment (bkawk) [inline API/tasks/getHashtags.js]: remove log

File: API/tasks/getHashtagItems.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+					"dateTime": 1508335855072,
+					"location": "u1557fpvm1hb",
+					"value": 12
+}
+
+				}]
+		;
+
+			return new Promise((resolve, reject) => {
+				console.log('HashtagsItems: ', returnObject);
```
Comment (bkawk) [inline API/tasks/getHashtagItems.js]: remove log

File: API/tasks.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      });
+    },
+
+    _setAvatar: function(queue, task) {
+      return Avatar._setAvatar(task.base64).then((res) => {
+        task.socket.emit('setAvatar', res);
+        queue(task, "remove");
+        return res;
+      }).catch((err) => {
+        console.log('set avatar ERR! ', err)
```
Comment (bkawk) [inline API/tasks.js]: call _errorLog

File: API/tasks.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      });
+    },
+
+    _getAvatar: function(queue, task) {
+      return Avatar._getAvatar(task.hash).then((res) => {
+        task.socket.emit('getAvatar', res);
+        queue(task, "remove");
+        return res;
+      }).catch((err) => {
+        console.log('get avatar ERR! ', err)
```
Comment (bkawk) [inline API/tasks.js]: call _errorLog

File: API/tasks.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+      });
+    },
+
+    _getHashtags: function(queue, task) {
+      return getHashtags._getHashtags(task.publicKey).then((res) => {
+        task.socket.emit('hashtagsChanged', res);
+        queue(task, "remove");
+        return res;
+      }).catch((err) => {
+        console.log('get hashtags ERR! ', err)
```
Comment (bkawk) [inline API/tasks.js]: call _errorLog

File: API/tasks.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+      });
+    },
+
+    _getGasPrice: function(queue, task) {
+      return getGasPrice._getGasPrice(task.publicKey).then((res) => {
+        task.socket.emit('gasPriceChanged', res);
+        queue(task, "remove");
+        return res;
+      }).catch((err) => {
+        console.log('get gasprice ERR! ', err)
```
Comment (bkawk) [inline API/tasks.js]: call _errorLog

File: API/tasks.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+      });
+    },
+
+    _getBalance: function(queue, task) {
+      return getBalance._getBalance(task.publicKey).then((res) => {
+        task.socket.emit('balanceChanged', res);
+        queue(task, "remove");
+        return res;
+      }).catch((err) => {
+        console.log('get balance ERR! ', err)
```
Comment (bkawk) [inline API/tasks.js]: call _errorLog

File: API/tasks.js [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Tasks
+     */
+    _getFx: function(queue, task) {
+      return getFx._getFx(task.publicKey).then((res) => {
+        task.socket.emit('fxChanged', res);
+        queue(task, "remove");
+        return res;
+      }).catch((err) => {
+        console.log('get fx ERR! ', err)
```
Comment (bkawk) [inline API/tasks.js]: call _errorLog

File: API/socket.js [hunk 17]
```diff
@@ -96,150 +138,146 @@ io.on('connection', function (socket) {
...
-            console.log('=========== GET FX ============');
-            resolve('getFx');
-        })
+        return taskList.reduce((chain, task) => {
+            return chain.then(() => task.toDo(_queue, task))
+                .then(val => console.log(val, Date.now()));
+        }, Promise.resolve()).then(() => {
+            taskInProgress = false
+        }).catch(() => {
+            taskInProgress = false
```
Comment (bkawk) [inline API/socket.js]: call _errorLog

File: API/socket.js [hunk 18]
```diff
@@ -96,150 +138,146 @@ io.on('connection', function (socket) {
...
-    * Tasks
-    */
-    function _getFx(data){
-        return new Promise((resolve, reject) => {
-            console.log('=========== GET FX ============');
-            resolve('getFx');
-        })
+        return taskList.reduce((chain, task) => {
+            return chain.then(() => task.toDo(_queue, task))
+                .then(val => console.log(val, Date.now()));
```
Comment (bkawk) [inline API/socket.js]: remove log

File: API/socket.js [hunk 19]
```diff
@@ -96,150 +138,146 @@ io.on('connection', function (socket) {
...
+    });
+
+    socket.on('hashtagItems', (data, response) => {
+        console.log('hashtagItems', data);
+        // Make promise, store deal and return txhash
+        return getHashtagItems._getHashtagItems(data).then((res) => {
+          socket.emit('hashtagItemsChanged', res);
+          response({ status: 200, res});
+        }).catch((err) => {
+          console.log('hashtagItems ERR! ', err)
```
Comment (bkawk) [inline API/socket.js]: call _errorLog

File: API/socket.js [hunk 20]
```diff
@@ -96,150 +138,146 @@ io.on('connection', function (socket) {
...
+        });
+    });
+    socket.on('createItem', (data, response) => {
+        console.log('createItem', data);
+        // Make promise, store deal and return txhash
+        return newItem._newItem(data).then((res) => {
+          socket.emit('newItemChanged', res);
+          response({ status: 200, res});
+        }).catch((err) => {
+          console.log('newItem ERR! ', err)
```
Comment (bkawk) [inline API/socket.js]: call _errorLog

File: API/socket.js [hunk 21]
```diff
@@ -8,78 +8,120 @@ const Web3 = require('web3');
...
     }
     const getPendingTransactions = {
         nextRun: 0,
         interval: 0,
-        toDo: _getPendingTransaction,
+        toDo: tasks._getPendingTransaction,
         publicKey: user.publicKey,
     }
+    const getFx = {
+            nextRun: 0,
```
Comment (bkawk) [inline API/socket.js]: Indenting

File: API/socket.js [hunk 22]
```diff
@@ -8,78 +8,120 @@ const Web3 = require('web3');
...
  * Connect
  */
-io.on('connection', function (socket) {
-    const user = {socketId: socket.id, publicKey: socket.handshake.query.publicKey};
+io.on('connection', function(socket) {
+
+    // the client should provide a publicKey argument
+    // if not - just diconnect
+    if (!socket.handshake.query.publicKey) {
+        console.log('no pubkey given.');
```
Comment (bkawk) [inline API/socket.js]: remove log

**Review by bkawk — CHANGES_REQUESTED** [review]

File: API/package.json (hunk 2 above)
Comment (bkawk) [reply API/package.json]: https://www.polymer-project.org/2.0/docs/tools/tests

File: API/tasks/getHashtagItems.js (hunk 10 above)
Comment (sponnet) [reply API/tasks/getHashtagItems.js]: done

File: API/tasks/getHashtags.js (hunk 9 above)
Comment (sponnet) [reply API/tasks/getHashtags.js]: done

File: Site/index.html (hunk 8 above)
Comment (sponnet) [reply Site/index.html]: done

File: API/tasks/getGasPrice.js (hunk 6 above)
Comment (sponnet) [reply API/tasks/getGasPrice.js]: done

File: API/package.json (hunk 2 above)
Comment (sponnet) [reply API/package.json]: the ```should``` module is part of the mocha test-suite , which is part of web-component-tester. It's the de-facto way of unit testing nodejs code.

File: API/tasks/getBalance.js (hunk 4 above)
Comment (sponnet) [reply API/tasks/getBalance.js]: done

File: API/package.json (hunk 1 above)
Comment (sponnet) [reply API/package.json]: removed !

File: API/tasks/Avatar.js (hunk 3 above)
Comment (sponnet) [reply API/tasks/Avatar.js]: @kingflurkel - can you do that ? TYS

File: API/package.json (hunk 1 above)
Comment (kingflurkel) [reply API/package.json]: done

File: API/tasks/Avatar.js (hunk 3 above)
Comment (kingflurkel) [reply API/tasks/Avatar.js]: gonna need a little help / clarification

File: API/socket.js (hunk 22 above)
Comment (kingflurkel) [reply API/socket.js]: done

File: API/socket.js (hunk 21 above)
Comment (kingflurkel) [reply API/socket.js]: done

File: API/socket.js (hunk 20 above)
Comment (kingflurkel) [reply API/socket.js]: done

File: API/socket.js (hunk 19 above)
Comment (kingflurkel) [reply API/socket.js]: done

File: API/socket.js (hunk 18 above)
Comment (kingflurkel) [reply API/socket.js]: done

File: API/socket.js (hunk 17 above)
Comment (kingflurkel) [reply API/socket.js]: done

**Review by bkawk — APPROVED** [review]

**Pull request merged**